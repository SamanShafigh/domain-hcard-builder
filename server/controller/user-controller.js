const UserService = require('../service/user-service')
const UserRepo = require('../repository/user-repository')
const { getKeyValue } = require('../util')

const USER_SSR_PATH = '../../client/dist/main.js';
const USER_SPA_PATH = '../../client/dist/_index.html';

class UserController {
  constructor(userService, config) {
    this.userService = userService;
    this.config = config;
  }

  index() {
    return async (ctx, next) => {
      ctx.view = {
        component: USER_SSR_PATH, 
        template:  USER_SPA_PATH,
        props: await this.userService.getUser(ctx.meta.userId)
      }
      await next();
    }
  }

  submit() {
    return async (ctx, next) => {
      await this.userService.submitUser(
        ctx.meta.userId, 
        ctx.request.body
      );

      ctx.view = {
        component: USER_SSR_PATH, 
        template:  USER_SPA_PATH,
        props: await this.userService.getUser(ctx.meta.userId)
      }
      await next();
    }
  }

  update() {
    return async (ctx, next) => {
      var { key, value } = getKeyValue(ctx.request.body);
      await this.userService.updateUser(
        ctx.meta.userId, 
        key,
        value
      );

      ctx.view = {
        response: `user field ${key} get updated to ${value}`
      }
      await next();
    }
  }
}

/**
 * makeUserController makes a 
 * 
 * @param {*} config 
 * @param {*} db 
 */
function makeUserController(config, db) {
  return new UserController(
    new UserService(
      new UserRepo(db)
    ),
    config
  )
}

module.exports = {
  UserController,
  makeUserController 
};
