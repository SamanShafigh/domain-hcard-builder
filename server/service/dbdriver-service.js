var MongoClient = require('mongodb').MongoClient;

async function connect({ dbUri, dbName }) {
  return new Promise((resolve, reject) => {
    MongoClient.connect(dbUri, {useNewUrlParser: true}, (err, client) => {
      if (err) {
        return reject(err)
      }
      var db = client.db(dbName)
      resolve(db)
    });
  })
}

module.exports = {
  connect
}