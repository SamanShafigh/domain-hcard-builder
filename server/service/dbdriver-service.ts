import { MongoClient } from 'mongodb';

export async function dbConnect({ dbUri, dbName }: any) {
  return new Promise((resolve: any, reject: any) => {
    MongoClient.connect(dbUri, {useNewUrlParser: true}, (err: any, client: any) => {
      if (err) {
        return reject(err)
      }
      var db = client.db(dbName)
      resolve(db)
    });
  })
}