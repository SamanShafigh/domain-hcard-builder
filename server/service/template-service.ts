const { createReadStream } = require('fs');
const React = require('react');
const ReactDom = require('react-dom/server')

interface Global {
  React: any;
}

declare var global: Global;
global.React = React;

function html(body: string) {
  return `
  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <meta http-equiv="X-UA-Compatible" content="IE=edge">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <title>Live hCard Preview</title>
  
      <link href="css/bootstrap.min.css" rel="stylesheet" >
      <link href="css/main.css" rel="stylesheet">
    </head>
  
    <body>
      ${body}
    </body>
  </html>
  `;
} 

function render(path: string, props: any) {
  const component = require(path).default;
  var body = ReactDom.renderToString(
    React.createElement(component, props)
  );

  return html(body);
}

function serve(path: string) {
  return createReadStream(`${__dirname}/${path}`)
}

module.exports = {
  render,
  serve
}