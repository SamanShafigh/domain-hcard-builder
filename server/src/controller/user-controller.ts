import UserService from '../service/user-service'
import UserRepo from '../repository/user-repository'
import { getKeyValue } from '../util'
import HCardT from '../type';

/**
 * UserController is to abstract the user related controllers
 */
class UserController implements HCardT.UserController {
  userService: HCardT.UserService;
  config: HCardT.Config;

  constructor(userService: HCardT.UserService, config: HCardT.Config) {
    this.userService = userService;
    this.config = config;
  }

  /** Get the view page */
  index() {
    return async (ctx: HCardT.CTX, next: HCardT.Next) => {
      ctx.view = {
        component: this.config.ssrPath, 
        template:  this.config.spaPath,
        props: await this.userService.getUser(ctx.meta.userId)
      }
      await next();
    }
  }

  /** Submit a user data endpoint */
  submit() {
    return async (ctx: HCardT.CTX, next: HCardT.Next) => {
      await this.userService.submitUser(
        ctx.meta.userId, 
        ctx.request.body
      );

      ctx.view = {
        component: this.config.ssrPath, 
        template:  this.config.spaPath,
        props: await this.userService.getUser(ctx.meta.userId)
      }
      await next();
    }
  }

  /** Update user data endpoint */
  update() {
    return async (ctx: HCardT.CTX, next: HCardT.Next) => {
      var { key, value }: HCardT.KeyValue = getKeyValue(ctx.request.body);
      await this.userService.updateUser(
        ctx.meta.userId, 
        key,
        value
      );

      ctx.view = {
        response: `user field ${key} get updated to ${value}`
      }
      await next();
    }
  }
}

/**
 * make makes a 
 * 
 * @param {*} config 
 * @param {*} db 
 */
export function make(config: HCardT.Config, dbDriver: any): HCardT.UserController {
  return new UserController(
    new UserService(dbDriver),
    config
  )
}
