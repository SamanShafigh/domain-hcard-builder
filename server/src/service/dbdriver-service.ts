import { MongoClient } from 'mongodb';
import HCardT from '../type';

/**
 * This is a default DB storage driver currently we support
 */
class MongoDbDriver implements HCardT.DbDriver {
  db: any
  constructor(db: any) {
    this.db = db;
  }

  /**
   * Find an entity
   * @param model 
   * @param query 
   * @param options 
   */
  async findOne(model: string, query: any, options?: any) {
    return await this.db
      .collection(model)
      .findOne(query, options);
  }

  /**
   * Save an entity
   * @param model 
   * @param query 
   * @param data 
   * @param options 
   */
  async save(model: string,query: any, data: any, options?: any) {
    await this.db
      .collection(model)
      .update(
        query, 
        {$set: data}, 
        Object.assign({upsert: true}, options)
      );
  }

  /**
   * Update an entity
   * @param model 
   * @param query 
   * @param data 
   * @param options 
   */
  async update(model: string, query: any, data: any, options?: any) {
    await this.db
      .collection(model)
      .update(
        query, 
        {$set: data}, 
        Object.assign({upsert: true}, options)
      );
  }  
}

/**
 * Make a Data storage driver, the default one is Mongo, but I can make it 
 * more advanced to return different type of drivers
 * 
 * @param dbUri 
 * @param dbName 
 */
export async function makeDbDriver(dbUri: string, dbName: string): Promise<any> {
  return new Promise((resolve: any, reject: any) => {
    // Just make a default mongoDB driver
    MongoClient.connect(dbUri, {useNewUrlParser: true}, (err: any, client: any) => {
      if (err) {
        return reject(err)
      }
      var db = client.db(dbName)
      var dbDriver = new MongoDbDriver(db);

      resolve(dbDriver)
    });
  })
}