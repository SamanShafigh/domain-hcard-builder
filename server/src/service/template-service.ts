import { createReadStream } from 'fs';
import * as React from 'react';
import * as ReactDom from 'react-dom/server';

interface Global { React: any; }
declare var global: Global;
global.React = React;

function html(body: string) {
  return `
  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <meta http-equiv="X-UA-Compatible" content="IE=edge">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <title>Live hCard Preview</title>
  
      <link href="css/bootstrap.min.css" rel="stylesheet" >
      <link href="css/main.css" rel="stylesheet">
    </head>
  
    <body>
      ${body}
    </body>
  </html>
  `;
} 

export function render(path: string, props: any) {
  const component = require(path).default;
  var body = ReactDom.renderToString(
    React.createElement(component, props)
  );

  return html(body);
}

export function serve(path: string) {
  return createReadStream(`${__dirname}/${path}`)
}
