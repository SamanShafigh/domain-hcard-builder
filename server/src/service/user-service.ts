import HCardT from '../type';
import makeUser from '../lib/model/user-model';

const MODEL_NAME = 'user'

/**
 * UserService is to abstract the main logic related to User
 */
class UserService implements HCardT.UserService {
  dbDriver: HCardT.DbDriver
  
  constructor(dbDriver: HCardT.DbDriver) {
    this.dbDriver = dbDriver;
  }

  /**
   * Get a user and if user does not exist returns null
   * @param userId 
   */
  async getUser(userId: string) {
    var result = await this.dbDriver
      .findOne(MODEL_NAME, { userId }, { _id: 0 });
    
    return result? makeUser(result.data) : null;
  }  

  /**
   * Submit a user data
   * @param userId 
   * @param data 
   */
  async submitUser(userId: string, data: any) {
    const user = makeUser(data);
    return await this.dbDriver
      .update(MODEL_NAME, { userId }, { data: user });
  }

  /**
   * Update a user parameter
   * @param userId 
   * @param key 
   * @param value 
   */
  async updateUser(userId: string, key: string, value: string) {
    var reference = `data.${key}`;

    return await this.dbDriver
      .update(MODEL_NAME, { userId }, { [reference]: value });
  }
}

export default UserService;
