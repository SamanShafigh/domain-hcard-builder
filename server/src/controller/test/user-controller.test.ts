import UserService from '../../service/user-service';
import * as dbdriver from '../../service/dbdriver-service';
import { make } from '../user-controller';
import { 
  mockConfig, 
  mockUser, 
  mockCtx, 
  mockCtxWithBody, 
  mockDbDriver } from '../../mock';

let userController;
jest.mock('../../service/dbdriver-service');

beforeAll(async () => {
  // Mocking my DB driver
  dbdriver.makeDbDriver.mockImplementationOnce(() => mockDbDriver)
  // Make my UserService
  const mockDriver = dbdriver.makeDbDriver();
  userController = make(mockConfig, mockDriver);
});

describe('Test user controller', () => {
  test('test get index page', async () => {
    var ctx = Object.assign({}, mockCtx);
    var middleware = await userController.index();
    await middleware(ctx, () => {});

    expect(ctx).toMatchSnapshot()
  });

  test('test submit api', async () => {
    var ctx = Object.assign({}, mockCtxWithBody);
    var middleware = await userController.submit();
    var result = await middleware(ctx, () => {});
    
    expect(ctx).toMatchSnapshot()
  });

  test('test update api', async () => {
    var ctx = Object.assign({}, mockCtxWithBody);
    var middleware = await userController.update();
    var result = await middleware(ctx, () => {});

    expect(ctx).toMatchSnapshot()
  });
});