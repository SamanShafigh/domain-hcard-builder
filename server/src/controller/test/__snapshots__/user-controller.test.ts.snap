// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test user controller test get index page 1`] = `
Object {
  "meta": Object {
    "noscript": true,
    "userId": "1234",
  },
  "view": Object {
    "component": "./dist/main.js",
    "props": User {
      "country": "country",
      "email": "email",
      "givenName": "givenName",
      "houseNumber": "houseNumber",
      "phone": "phone",
      "postcode": "postcode",
      "state": "state",
      "street": "street",
      "suburb": "suburb",
      "surname": "surname",
    },
    "template": "./dist/_index.html",
  },
}
`;

exports[`Test user controller test submit api 1`] = `
Object {
  "meta": Object {
    "noscript": true,
    "userId": "1234",
  },
  "request": Object {
    "body": Object {
      "country": "country",
      "email": "email",
      "givenName": "givenName",
      "houseNumber": "houseNumber",
      "phone": "phone",
      "postcode": "postcode",
      "state": "state",
      "street": "street",
      "suburb": "suburb",
      "surname": "surname",
    },
  },
  "view": Object {
    "component": "./dist/main.js",
    "props": User {
      "country": "country",
      "email": "email",
      "givenName": "givenName",
      "houseNumber": "houseNumber",
      "phone": "phone",
      "postcode": "postcode",
      "state": "state",
      "street": "street",
      "suburb": "suburb",
      "surname": "surname",
    },
    "template": "./dist/_index.html",
  },
}
`;

exports[`Test user controller test update api 1`] = `
Object {
  "meta": Object {
    "noscript": true,
    "userId": "1234",
  },
  "request": Object {
    "body": Object {
      "country": "country",
      "email": "email",
      "givenName": "givenName",
      "houseNumber": "houseNumber",
      "phone": "phone",
      "postcode": "postcode",
      "state": "state",
      "street": "street",
      "suburb": "suburb",
      "surname": "surname",
    },
  },
  "view": Object {
    "response": "user field givenName get updated to givenName",
  },
}
`;
